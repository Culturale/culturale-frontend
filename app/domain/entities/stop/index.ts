export * from './stop';
export * from './stop.interface';
