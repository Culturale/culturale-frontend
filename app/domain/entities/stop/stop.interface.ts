/**
 * Representation of a stop in the driver route
 *
 * @property stopId -> UUID to identify the stop
 * @property address -> Address of the stop
 * @property sequence -> Index of the stop inside the route
 * @property clientRadius -> A perimeter set for client side delivery
 * @property retailerRadius -> A perimeter set for retailer side delivery
 * @property completed -> If the stop is completed is set as true
 * @property eta -> Time window in which the driver is expected to arrive
 */

export interface IStop {
  stopId: string;
  address: string;
  sequence?: number;
  clientRadius?: number;
  retailerRadius?: number;
  completed?: boolean;
  geoHash: string;

  /**
   * Returns if the order is completed
   * @readonly
   */
  readonly isCompleted: boolean;

  /**
   * Sets completed as true
   * @public
   */
  completeStop: () => void;

  /**
   * Sets completed as false
   * @public
   */
  uncompleteStop: () => void;

  /**
   * Changes the perimeter of the stop
   * @public
   */
  updatePerimeter: (clientRadius: number, retailerRadius: number) => void;

  /**
   * Returns the sequence number of the stop
   * @public
   */
  readonly sequenceNumber: number;
}
