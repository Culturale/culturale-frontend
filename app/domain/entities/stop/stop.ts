import { action, computed, makeObservable, observable } from 'mobx';

import type { IStop } from './stop.interface';

export class Stop implements IStop {
  public stopId: string;
  public address: string;
  public clientRadius?: number;
  public completed?: boolean;
  public retailerRadius?: number;
  public geoHash: string;

  constructor(stopId: string, address?: string, completed?: boolean) {
    this.stopId = stopId;
    this.completed = completed;
    this.address = address;

    makeObservable(this, {
      completeStop: action,
      completed: observable,
      isCompleted: computed,
      stopId: observable,
      uncompleteStop: action,
      updatePerimeter: action
    });
  }

  sequence?: number;

  public completeStop() {
    this.completed = true;
  }

  public uncompleteStop() {
    this.completed = false;
  }

  get isCompleted(): boolean {
    return this.completed;
  }

  public updatePerimeter(clientRadius: number, retailerRadius: number) {
    this.clientRadius = clientRadius;
    this.retailerRadius = retailerRadius;
  }

  public get sequenceNumber(): number {
    return 0;
  }
}
