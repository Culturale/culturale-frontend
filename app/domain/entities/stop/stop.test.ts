import { Stop } from './stop';
import type { IStop } from './stop.interface';

describe('Stop', function () {
  let instance: IStop;

  beforeEach(function () {
    instance = new Stop('test-id');
  });

  it('can be created', function () {
    expect(instance).toMatchSnapshot();
  });

  describe('completeStop', function () {
    it('sets the stop as completed', function () {
      instance.completeStop();

      expect(instance.completed).toBeTruthy();
    });
  });

  describe('uncompleteStop', function () {
    it('set the stop as not completed', function () {
      instance.completeStop();
      instance.uncompleteStop();

      expect(instance.completed).toBeFalsy();
    });
  });

  describe('get sequenceNumber', function () {
    it('returns 0', function () {
      expect(instance.sequenceNumber).toBe(0);
    });
  });
});
