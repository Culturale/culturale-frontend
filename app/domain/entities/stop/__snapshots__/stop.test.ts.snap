// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stop can be created 1`] = `
Stop {
  "address": undefined,
  "clientRadius": undefined,
  "completed": undefined,
  "geoHash": undefined,
  "retailerRadius": undefined,
  "sequence": undefined,
  "stopId": "test-id",
}
`;
